File: application/chann_app/services/deal_value.py
"""What a deal is worth, given a deal dict (round 21C).

The Data tier's `repositories/deal_value.py` is the same rule in SQL. This
one exists because the Application tier holds deals as dicts from the
DataClient, and three places were each deciding for themselves what a deal
was worth — two of them differently from the pipeline card.

`amount` is nullable. NULL and "" mean nobody typed a value; 0 means
somebody typed zero, and a typed zero is an answer.
"""
from __future__ import annotations

from decimal import Decimal, InvalidOperation


def _decimal(value) -> Decimal:
    try:
        return Decimal(str(value if value not in (None, "") else 0))
    except (InvalidOperation, ValueError):
        return Decimal("0")


def deal_value(deal: dict) -> Decimal:
    """The typed amount when there is one, otherwise the line items."""
    typed = deal.get("amount")
    if typed is not None and str(typed).strip() != "":
        return _decimal(typed)
    total = Decimal("0")
    for row in deal.get("products") or []:
        total += _decimal(row.get("quoted_unit_price")) * int(row.get("qty") or 0)
    return total

File: presentation/app/liff/sales/_deal-value.ts
/** What a deal is worth — the screens' copy of the one rule
 *  (`application/chann_app/services/deal_value.py`, and the SQL in
 *  `data/chann_data/repositories/deal_value.py`): the amount the
 *  salesperson typed when there is one — a typed 0 is a typed value —
 *  otherwise the line items. Pre-VAT, like every deal value on the
 *  platform (ruling 23). One copy, so the deal list, the deal page and the
 *  chat side panel cannot each decide for themselves (final review I4). */
export type DealValueInput = {
  amount?: string | number | null;
  products?: { qty?: number | string | null; quoted_unit_price?: string | number | null }[] | null;
};

export function dealValue(deal: DealValueInput): number {
  if (deal.amount != null && deal.amount !== "") return Number(deal.amount) || 0;
  return (deal.products ?? []).reduce(
    (sum, p) => sum + Number(p.qty ?? 0) * Number(p.quoted_unit_price ?? 0), 0,
  );
}

/** Whether the deal carries any value at all — neither a typed amount nor
 *  a line means "nothing to show", which is not the same as 0. */
export function hasDealValue(deal: DealValueInput): boolean {
  return (deal.amount != null && deal.amount !== "") || (deal.products ?? []).length > 0;
}
